Process aliases

mat-hek opened the pull request:
TODO:

- [x] support aliases in `spawn_opt`
- [x] reply_demonitor
- [x] C docs
- [x] erlang docs
- [x] process reference serialisation


These changes are made under both the "Apache 2.0" and the "GNU Lesser General
Public License 2.1 or later" license terms (dual license).

SPDX-License-Identifier: Apache-2.0 OR LGPL-2.1-or-later

mat-hek committed:
Process aliases

Signed-off-by: Mateusz Front <mateusz.front@swmansion.com>

bettio commented on src/libAtomVM/term.h:
```diff
@@ -148,6 +148,8 @@ extern "C" {
 #define BOXED_FUN_SIZE 3
 #define FLOAT_SIZE (sizeof(float_term_t) / sizeof(term) + 1)
 #define REF_SIZE ((int) ((sizeof(uint64_t) / sizeof(term)) + 1))
+#define TERM_BOXED_PID_REF_SIZE (REF_SIZE + 1)
```
For consistency with the work we are doing lately I suggest renaming all `PID_REF` and `pid_reference` to `PROCESS_REF` and `process_reference`

bettio commented on tests/erlang_tests/test_monitor.erl:
```diff
@@ -34,6 +34,10 @@ start() ->
+    ok = test_alias(),
+    ok = test_monitor_alias(),
+    ok = test_monitor_alias_explicit_unalias(),
+    ok = test_monitor_down_alias(),
```
We should also test unhappy paths, such as:
1. double/triple unalias
2. alias a dead process
3. unalias a dead process

In addition multiple aliases should be tested,  this is legit but is should be properly tested.

bettio commented on src/libAtomVM/nifs.c:
```diff
@@ -4216,17 +4236,49 @@ static term nif_erlang_memory(Context *ctx, int argc, term argv[])
         RAISE_ERROR(BADARG_ATOM);
     }
 
+    while (!term_is_nil(options)) {
```
We should use `term_is_nonempty_list` to avoid issues with non proper lists.

bettio commented on src/libAtomVM/context.h:
```diff
@@ -172,6 +172,13 @@ enum ContextMonitorType
+enum ContextMonitorAliasType
+{
+    CONTEXT_MONITOR_ALIAS_EXPLICIT_UNALIAS,
+    CONTEXT_MONITOR_ALIAS_DEMONITOR,
```
Let's use PascalCase. See AVMCCS-N004 rule in our coding style guide.
(There is still some old code that has to be migrated to updated style).

bettio reviewed (changes requested):
Looks mostly good: there is only one change related to `IS_NULL_PTR` that is required. I suggest also rebasing on top of main, since main has a lot of fixes for flaky tests.

pguyot commented on src/libAtomVM/context.c:
```diff
@@ -774,15 +775,15 @@ static struct Monitor *context_monitors_handle_terminate(Context *ctx)
                 }
                 // Prepare the message on ctx's heap which will be freed afterwards.
-                term ref = term_from_ref_ticks(monitored_monitor->ref_ticks, &ctx->heap);
+                term ref = term_make_process_reference(target->process_id, monitored_monitor->ref_ticks, &ctx->heap);
```
This is not what OTP does and there probably is a good reason.
Regular monitors are smaller than aliases.
```
7> Pid = spawn(fun() -> receive ok -> ok end end).
<0.94.0>
8> monitor(process, Pid).
#Ref<0.3008598808.1200095248.57512>
9> monitor(process, Pid, [{alias, explicit_unalias}]).
#Ref<0.0.11651.3008598808.1200160784.57533>
10> make_ref().
#Ref<0.3008598808.1200095248.57553>
```

pguyot commented on src/libAtomVM/term.h:
```diff
@@ -1959,6 +1983,50 @@ static inline uint64_t term_to_ref_ticks(term rt)
+    boxed_value[0] = TERM_BOXED_PROCESS_REF_HEADER;
+
+#if TERM_BYTES == 4
+    boxed_value[1] = (ref_ticks >> 4);
```
This was fixed for regular refs, we probably want ref_ticks >> 32

pguyot commented on tests/erlang_tests/test_binary_to_term.erl:
```diff
@@ -1136,23 +1137,34 @@ test_encode_resource(OTPVersion) ->
     end,
     Term.
 
+is_atomvm_or_otp_version_at_least(OTPVersion) ->
```
`get_otp_version/0` already exists in this module.

Also there is a poor usage (fixed in #2025 )
```
    OTPVersion = get_otp_version(),
    if
        OTPVersion =:= atomvm orelse OTPVersion >= 26 ->
            B == erlang:term_to_binary(A);
        true ->
            true
    end.
```

be sure to rather do:
```
    OTPVersion = get_otp_version(),
    if
        OTPVersion >= 26 ->
            B == erlang:term_to_binary(A);
        true ->
            true
    end.
```
as atoms sort higher than integers.

pguyot reviewed (commented):
My main concern so far is that monitors are now extended to 2 or 3 words instead of 1 or 2. This may break some code that took monitors and passed it around as references, as the term_to_ref_ticks/term_from_ref_ticks is a common pattern. A similar breakage existed with resources being larger references if one tries to pass a resource where a 1/2 words ref is expected, but it is less likely as resources were not references so far. This may be the crash cause of esp32 tests.

port.erl
```
call(Port, Message, Timeout) ->
    MonitorRef = monitor(port, Port),
    Port ! {'$call', {self(), MonitorRef}, Message},
    Result =
        receive
            {'DOWN', MonitorRef, port, Port, normal} ->
                {error, noproc};
            {'DOWN', MonitorRef, port, Port, Reason} ->
                {error, Reason};
            out_of_memory ->
                out_of_memory;
            {MonitorRef, Ret} ->
                Ret
        after Timeout ->
            {error, timeout}
        end,
    demonitor(MonitorRef, [flush]),
    Result.
```


uart_driver.c:
```
static void uart_driver_do_read(Context *ctx, GenMessage gen_message)
{
    GlobalContext *glb = ctx->global;
    struct UARTData *uart_data = ctx->platform_data;
    term pid = gen_message.pid;
    term ref = gen_message.ref;
    uint64_t ref_ticks = term_to_ref_ticks(ref);
```

mat-hek commented:
@pguyot the reason for the STM32 tests failing was that the size of a regular reference is actually 3 terms there, which means process reference takes 4 terms, and it probably conflicts with resource reference size, which is always 4 terms. I changed the process reference size to 5, but that's not a perfect solution :P So we need to come up with another way of distinguishing references - do you have anything in mind?

I also added `RefData` as a more universal representation of references than `ref_ticks`. WYDT? Would it make sense for it to support resource references too?

mat-hek committed:
Add RefData, fix STM32 tests

Signed-off-by: Mateusz Front <mateusz.front@swmansion.com>

mat-hek committed:
Make monitors keep RefData

Signed-off-by: Mateusz Front <mateusz.front@swmansion.com>

mat-hek committed:
Make non-alias monitors short refs

Signed-off-by: Mateusz Front <mateusz.front@swmansion.com>

mat-hek committed:
Change reference sizes

Signed-off-by: Mateusz Front <mateusz.front@swmansion.com>

mat-hek committed:
Rename REF_SIZE -> SHORT_REF_SIZE

Signed-off-by: Mateusz Front <mateusz.front@swmansion.com>

mat-hek committed:
CR

Signed-off-by: Mateusz Front <mateusz.front@swmansion.com>

mat-hek committed:
Support for external and resource refs in RefData

Signed-off-by: Mateusz Front <mateusz.front@swmansion.com>

mat-hek committed:
CR fixes

Signed-off-by: Mateusz Front <mateusz.front@swmansion.com>

mat-hek committed:
rename ref sizes

Signed-off-by: Mateusz Front <mateusz.front@swmansion.com>

mat-hek committed:
fix formatting

Signed-off-by: Mateusz Front <mateusz.front@swmansion.com>

mat-hek commented on src/libAtomVM/term.h:
```diff
@@ -129,7 +129,14 @@ extern "C" {
+#define TERM_BOXED_REFERENCE_RESOURCE_SIZE 5
+#endif
 #define TERM_BOXED_REFERENCE_RESOURCE_HEADER (((TERM_BOXED_REFERENCE_RESOURCE_SIZE - 1) << 6) | TERM_BOXED_REF)
 #define TERM_BOXED_RESOURCE_SIZE TERM_BOXED_REFERENCE_RESOURCE_SIZE
```
I'd prefer sticking to the current option for now if it's acceptable, and change it in a separate PR

bettio commented:
Which clang-format version are you using?
clang-format has slightly changed style (for better) in recent versions and now this is formatted in a different way:
```
-            ref_data = (RefData) {
+            ref_data = (RefData){
                 .type = RefTypeProcess,
                 .process = { .ref_ticks = globalcontext_get_ref_ticks(ctx->global), .process_id = ctx->process_id }
             };
@@ -4389,7 +4389,7 @@
     RefData ref_data;
     struct Monitor *alias_monitor = NULL;
     if (is_alias) {
-        ref_data = (RefData) {
+        ref_data = (RefData){
```

This causes some friction, but I'm not sure yet if just updating clang-format is the right thing to do.

bettio reviewed (commented):
This has to be changed:

```
/home/runner/work/AtomVM/AtomVM/src/libAtomVM/term.h:3148:12: error: ‘ref_data’ may be used uninitialized [-Werror=maybe-uninitialized]
 3148 |     return ref_data;
      |            ^~~~~~~~
/home/runner/work/AtomVM/AtomVM/src/libAtomVM/term.h: In function ‘process_console_mailbox’:
/home/runner/work/AtomVM/AtomVM/src/libAtomVM/term.h:3129:13: note: ‘ref_data’ declared here
 3129 |     RefData ref_data;
      |             ^~~~~~~~
In function ‘term_to_ref_data’,
    inlined from ‘process_console_message’ at /home/runner/work/AtomVM/AtomVM/src/libAtomVM/nifs.c:1163:28,
    inlined from ‘process_console_mailbox’ at /home/runner/work/AtomVM/AtomVM/src/libAtomVM/nifs.c:1229:18:
/home/runner/work/AtomVM/AtomVM/src/libAtomVM/term.h:3148:12: error: ‘ref_data’ may be used uninitialized [-Werror=maybe-uninitialized]
 3148 |     return ref_data;
      |            ^~~~~~~~
/home/runner/work/AtomVM/AtomVM/src/libAtomVM/term.h: In function ‘process_console_mailbox’:
/home/runner/work/AtomVM/AtomVM/src/libAtomVM/term.h:3129:13: note: ‘ref_data’ declared here
 3129 |     RefData ref_data;
```

mat-hek committed:
CR

Signed-off-by: Mateusz Front <mateusz.front@swmansion.com>

mat-hek commented:
@bettio I have `clang-format version 18.1.8`. I disabled the `SpaceAfterCStyleCast` rule, let's see if it helps.

pguyot commented on src/libAtomVM/term.c:
```diff
@@ -405,6 +405,12 @@ int term_funprint(PrinterFun *fun, term t, const GlobalContext *global)
+        uint64_t ref_ticks = term_to_ref_ticks(t);
+
+        // Update also REF_AS_CSTRING_LEN when changing this format string
+        return fun->print(fun, "#Ref<%" PRId32 ".%" PRIu32 ".%" PRIu32 ">", process_id, (uint32_t) (ref_ticks >> 32), (uint32_t) ref_ticks);
```
Good catch! We should update tests as well to ensure comparison works. I added some tests when I made resources references.

mat-hek committed:
fix reference_process_pid_offset

Signed-off-by: Mateusz Front <mateusz.front@swmansion.com>

mat-hek committed:
fix parse_monitor_opts

Signed-off-by: Mateusz Front <mateusz.front@swmansion.com>

mat-hek committed:
Revert ref_data back to ref_ticks in drivers

Signed-off-by: Mateusz Front <mateusz.front@swmansion.com>

mat-hek committed:
REF_SIZE -> TERM_BOXED_SHORT_REFERENCE_SIZE in drivers

Signed-off-by: Mateusz Front <mateusz.front@swmansion.com>

mat-hek committed:
Add aliases to test_refs_ordering

Signed-off-by: Mateusz Front <mateusz.front@swmansion.com>

mat-hek committed:
Make comparing process refs compliant with the BEAM

Signed-off-by: Mateusz Front <mateusz.front@swmansion.com>

bettio commented on src/libAtomVM/opcodesswitch.h:
```diff
@@ -2777,12 +2777,28 @@ HOT_FUNC int scheduler_entry_point(GlobalContext *glb)
+                            TRACE_SEND(ctx, x_regs[0], x_regs[1]);
+                            globalcontext_send_message(ctx->global, local_process_id, x_regs[1]);
+                        } else if (term_is_process_reference(recipient_term)) {
+                            int32_t local_process_id = term_process_ref_to_process_id(recipient_term);
```
This same code block is duplicated in `jit.c`. I suggest creating a helper function to avoid code duplication, such as: globalcontext_send_message_to_alias(GlobalContext *, term alias, term message);

bettio commented on src/libAtomVM/jit.c:
```diff
@@ -719,7 +720,26 @@ static bool jit_send(Context *ctx, JITState *jit_state)
         globalcontext_send_message(ctx->global, local_process_id, ctx->x[1]);
         ctx->x[0] = ctx->x[1];
+    } else if (term_is_process_reference(recipient_term)) {
+        int32_t process_id = term_process_ref_to_process_id(recipient_term);
```
See my comment about duplicated code in opcodesswitch.h.

bettio commented on src/libAtomVM/nifs.c:
```diff
@@ -1150,7 +1160,7 @@ static NativeHandlerResult process_console_message(Context *ctx, term msg)
         term ref = term_get_tuple_element(msg, 2);
         term req = term_get_tuple_element(msg, 3);
-        uint64_t ref_ticks = term_to_ref_ticks(ref);
+        RefData ref_data = term_to_ref_data(ref);
```
This is a port as well, same thing I mentioned in my previous PR also applies here.

bettio commented on src/libAtomVM/nifs.c:
```diff
@@ -1160,7 +1170,7 @@ static NativeHandlerResult process_console_message(Context *ctx, term msg)
                 free(str);
 
-                term refcopy = term_from_ref_ticks(ref_ticks, &ctx->heap);
+                term refcopy = term_from_ref_data(ref_data, &ctx->heap);
```
This is a port as well, same thing I mentioned in my previous PR also applies here.

bettio commented on src/libAtomVM/nifs.c:
```diff
@@ -6343,6 +6459,45 @@ static term nif_erlang_list_to_bitstring_1(Context *ctx, int argc, term argv[])
+    RefData ref_data = {
+        .type = RefTypeProcess,
+        .process = { .ref_ticks = globalcontext_get_ref_ticks(ctx->global), .process_id = ctx->process_id }
+    };
```
As a feedback I suggest a shorter lines here:
```c
    RefData ref_data = {
        .type = RefTypeProcess,
        .process = {
            .ref_ticks = globalcontext_get_ref_ticks(ctx->global),
            .process_id = ctx->process_id
        }
    };
```

bettio reviewed (changes requested):
Review is still in progress, but meanwhile I submit comments for what I reviewed.

mat-hek committed:
CR

Signed-off-by: Mateusz Front <mateusz.front@swmansion.com>

bettio commented on src/libAtomVM/nifs.c:
```diff
@@ -1536,6 +1613,20 @@ static term nif_erlang_send_2(Context *ctx, int argc, term argv[])
 
         globalcontext_send_message(glb, local_process_id, argv[1]);
 
+    } else if (term_is_process_reference(target)) {
```
This is also a duplicate.

mat-hek commented on src/libAtomVM/context.h:
```diff
@@ -196,18 +203,25 @@ struct LinkLocalMonitor
 {
     struct Monitor monitor;
-    uint64_t ref_ticks;
+    RefData ref_data;
```
In this particular case, RefData is needed, because when making a term out of it, we need to know whether to create a process ref or a regular ref. If we need a process ref, then we also need a process id.

mat-hek committed:
CR

Signed-off-by: Mateusz Front <mateusz.front@swmansion.com>

mat-hek committed:
fix leaked git conflict

Signed-off-by: Mateusz Front <mateusz.front@swmansion.com>

mat-hek committed:
Adjust freeing alias monitors

Signed-off-by: Mateusz Front <mateusz.front@swmansion.com>

mat-hek committed:
Use short reference for monitors of dead processes

Signed-off-by: Mateusz Front <mateusz.front@swmansion.com>
